engine/src/main/scala/cromwell/engine/workflow/ShadowWorkflowManagerActor.scala

```diff
@@ -0,0 +1,255 @@
+     Watched transitions
+     */
+    case Event(Transition(workflowActor, _, toState: ShadowWorkflowActorTerminalState), data) =>
+      log.info(workflowActor.path.name + "has gone terminal")
+      stay using data.without(workflowActor)
+  }
+
+  when (Aborting) {
+    case Event(Transition(workflowActor, _, toState: WorkflowState), data) if toState.isTerminal =>
+      // Remove this terminal actor from the workflowStore and log a progress message.
+      val updatedData = data.without(workflowActor)
+      logger.info(s"$tag: Waiting for all workflows to abort (${updatedData.workflows.size} remaining).")
+      // If there are no more workflows to abort we're done, otherwise just stay in the current state.
+      (if (updatedData.workflows.isEmpty) goto(Done) else stay()) using updatedData
+  }
+
+  when (Done) { FSM.NullFunction }
+
+  whenUnhandled {
+    // Uninteresting transition and current state notifications.
+    case Event((Transition(_, _, _) | CurrentState(_, _)), _) => stay()
+    // Anything else certainly IS interesting:
+    case Event(unhandled, data) =>
+      log.warning(s"$tag: Unhandled message: $unhandled")
+      stay()
+  }
+
+  onTransition {
+    case _ -> Done =>
+      logger.info(s"$tag: All workflows aborted.")
+      donePromise.trySuccess(())
+    case fromState -> toState =>
+      logger.info(s"$tag transitioning from $fromState to $toState")
+  }
+
+
+  /** Submit the workflow and return an updated copy of the state data reflecting the addition of a
+    * Workflow ID -> WorkflowActorRef entry.
+    */
+  private def submitWorkflow(source: WorkflowSourceFiles, replyTo: Option[ActorRef], id: Option[WorkflowId]): WorkflowIdToActorRef = {
+    val workflowId: WorkflowId = id.getOrElse(WorkflowId.randomId())
+    logger.info(s"$tag submitWorkflow input id = $id, effective id = $workflowId")
+    val isRestart = id.isDefined
+
+    val startMode = if (isRestart) RestartExistingWorkflow else StartNewWorkflow
+    val wfActor = context.actorOf(ShadowWorkflowActor.props(workflowId, startMode, source, config), name = s"WorkflowActor-$workflowId")
+    replyTo.foreach { _ ! WorkflowManagerSubmitSuccess(id = workflowId) }
+    wfActor ! SubscribeTransitionCallBack(self)
+    wfActor ! StartWorkflowCommand
+    logger.info(s"Successfuly started ${wfActor.path} for Workflow $workflowId")
+    WorkflowIdToActorRef(workflowId, wfActor)
+  }
+
+  private def restartWorkflow(restartableWorkflow: WorkflowDescriptor): WorkflowIdToActorRef = {
+    logger.info("Invoking restartableWorkflow on " + restartableWorkflow.id.shortString)
+    submitWorkflow(restartableWorkflow.sourceFiles, replyTo = None, Option(restartableWorkflow.id))
+  }
+
+  private def restartIncompleteWorkflows(): Unit = {
+    def logRestarts(restartableWorkflows: Traversable[WorkflowDescriptor]): Unit = {
+      val num = restartableWorkflows.size
+      val displayNum = if (num == 0) "no" else num.toString
+      val plural = if (num == 1) "" else "s"
```

Comment:
I thought we had an enhanced something or other floating around to do the pluralization somewhere. There's also something in scalaz but IIRC it is kinda hokey